### PR TITLE
Make softforked consensus changes unconditionally activated in Zcash

### LINKED ISSUE #968: Enforce strict DER for ECDSA signatures and compressed ECDSA public keys
Bitcoin started enforcing "strict" DER ECDSA signatures as a *standard* rule [in v0.8.0](https://bitcointalk.org/index.php?topic=106461.msg3932995#msg3932995). It does not enforce this or canonicity of ECDSA public keys as an acceptance rule. I propose that Zcash should enforce these as acceptance rules from the start.

### PR DESCRIPTION
**Pull request by ebfull**
Closes #968

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -2024,33 +2024,14 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
-
-    unsigned int flags = fStrictPayToScriptHash ? SCRIPT_VERIFY_P2SH : SCRIPT_VERIFY_NONE;
+    BOOST_FOREACH(const CTransaction& tx, block.vtx) {
+        const CCoins* coins = view.AccessCoins(tx.GetHash());
+        if (coins && !coins->IsPruned())
+            return state.DoS(100, error("ConnectBlock(): tried to overwrite transaction"),
+                             REJECT_INVALID, "bad-txns-BIP30");
+    }
+
+    unsigned int flags = SCRIPT_VERIFY_P2SH;
```
Comment (defuse) [inline src/main.cpp]: Add a comment referencing BIP 16

**Comment by defuse** [comment]
Not sure if this was ready for review, but ACK

**Comment by ebfull** [comment]
I've changed the scope of this PR as per our discussion at the engineering meeting; in addition to before, this will also force blocks to be version 4 and above, unconditionally instituting the remaining softforks in upstream Bitcoin.

**Comment by ebfull** [comment]
Rebased and pushed the remaining changes.

**Comment by defuse** [comment]
ACK

**Comment by ebfull** [comment]
@zkbot r+

**Comment by ebfull** [comment]
@zkbot r-

**Comment by ebfull** [comment]
@zkbot retry

**Comment by ebfull** [comment]
@zkbot r+

**Comment by ebfull** [comment]
@zkbot force retry

**Comment by daira** [comment]
Post-hoc ACK.

**Pull request merged**